WalletWasabi/Hwi/HwiProcessManager.cs

```diff
@@ -0,0 +1,244 @@
+				//	string err = null;
+				//	try
+				//	{
+				//		JObject json = jToken as JObject;
+				//		err = json.Value<string>("error");
+				//	}
+				//	catch (Exception ex)
+				//	{
+				//		Logger.LogTrace(ex, nameof(HwiProcessManager));
+				//	}
+				//	if (err != null)
+				//	{
+				//		throw new IOException(err);
+				//	}
+
+				//	return jToken;
+				//}
+			}
+		}
+
+		public static async Task EnsureHwiInstalledAsync(string dataDir, Network network)
+		{
+			Network = network;
+
+			var fullBaseDirectory = Path.GetFullPath(AppContext.BaseDirectory);
+			if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				if (!fullBaseDirectory.StartsWith('/'))
+				{
+					fullBaseDirectory.Insert(0, "/");
+				}
+			}
+
+			var hwiDir = Path.Combine(dataDir, "hwi");
+
+			string hwiPath;
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				hwiPath = $@"{hwiDir}\hwi.exe";
+			}
+			else // Linux or OSX
+			{
+				hwiPath = $@"{hwiDir}/hwi";
+			}
+
+			if (!File.Exists(hwiPath))
+			{
+				Logger.LogInfo($"HWI instance NOT found at {hwiPath}. Attempting to acquire it...", nameof(HwiProcessManager));
+				await InstallHwiAsync(fullBaseDirectory, hwiDir);
+			}
+			else if (new FileInfo(hwiPath).CreationTimeUtc < new DateTime(2019, 04, 13, 0, 0, 0, 0, DateTimeKind.Utc))
+			{
+				Logger.LogInfo($"Updating HWI...", nameof(HwiProcessManager));
+
+				string backupHwiDir = $"{hwiDir}_backup";
+				if (Directory.Exists(backupHwiDir))
+				{
+					Directory.Delete(backupHwiDir, true);
+				}
+				Directory.Move(hwiDir, backupHwiDir);
+
+				await InstallHwiAsync(fullBaseDirectory, hwiDir);
+			}
+			else
+			{
+				Logger.LogInfo($"HWI instance found at {hwiPath}.", nameof(HwiProcessManager));
+			}
+
+			HwiPath = hwiPath;
+		}
+
+		private static async Task InstallHwiAsync(string fullBaseDirectory, string hwiDir)
+		{
+			string hwiSoftwareDir = Path.Combine(fullBaseDirectory, "Hwi", "Software");
+
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				string hwiWinZip = Path.Combine(hwiSoftwareDir, "hwi-win64.zip");
+				await IoHelpers.BetterExtractZipToDirectoryAsync(hwiWinZip, hwiDir);
+				Logger.LogInfo($"Extracted {hwiWinZip} to {hwiDir}.", nameof(HwiProcessManager));
+			}
+			else // Linux or OSX
+			{
+				if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+				{
+					string hwiLinuxZip = Path.Combine(hwiSoftwareDir, "hwi-linux64.zip");
+					await IoHelpers.BetterExtractZipToDirectoryAsync(hwiLinuxZip, hwiDir);
+					Logger.LogInfo($"Extracted {hwiLinuxZip} to {hwiDir}.", nameof(HwiProcessManager));
+				}
+				else // OSX
+				{
+					string hwiOsxZip = Path.Combine(hwiSoftwareDir, "hwi-osx64.zip");
+					await IoHelpers.BetterExtractZipToDirectoryAsync(hwiOsxZip, hwiDir);
+					Logger.LogInfo($"Extracted {hwiOsxZip} to {hwiDir}.", nameof(HwiProcessManager));
+				}
+
+				// Make sure there's sufficient permission.
+				string chmodHwiDirCmd = $"chmod -R 777 {hwiDir}";
```

Comment:
I don't think it is a good idea to give full access to everyone to a process that handles money, in fact I think from a privacy point of view is risky too (anyone can extract the xpub, for example). 

I think we could give all the permission (wrx) to the logged user, who is the one running Wasabi and that is the owner of the hwi cli tool. Also give permissions to read and execute to the user's group and give no permissions at all to others.

```suggestion
				string chmodHwiDirCmd = $"chmod -R 750 {hwiDir}";
```

Reply:
ACK